guidance/guidance_los/guidance_los/los_guidance_action_server.py

```diff
@@ -0,0 +1,362 @@
+        self.log_publisher = self.create_publisher(
+            String, '/guidance_action_server/logs', 10
+        )
+
+        # Subscriber for vehicle state updates
+        self.create_subscription(
+            Odometry,
+            '/nucleus/odom',
+            self.odom_callback,
+            10,
+            callback_group=self.timer_cb_group,
+        )
+
+        # Publisher for visualizing waypoints in GUI
+        self.waypoint_list_pub = self.create_publisher(
+            WaypointList, '/guidance/waypoint_list', 10
+        )
+
+        # Action server for handling navigation requests
+        self._action_server = ActionServer(
+            self,
+            NavigateWaypoints,
+            'navigate_waypoints',
+            execute_callback=self.execute_callback,
+            callback_group=self.action_cb_group,
+        )
+
+        # State variables initialization
+        self.state: State = State()  # Current vehicle state
+        self.waypoints = []  # List of waypoints to navigate
+        self.current_waypoint_index = 0  # Index of current target waypoint
+        self.goal_handle = None  # Handle for current navigation goal
+        self.update_period = 0.1  # Guidance update period in seconds
+
+        self.get_logger().info('Guidance Action Server initialized')
+
+    def publish_log(self, message: str):
+        """Publish debug log message to ROS topic and node logger.
+
+        Args:
+            message (str): Message to be logged
+        """
+        msg = String()
+        msg.data = message
+        # self.log_publisher.publish(msg)
+        self.get_logger().info(message)
+
+    def publish_waypoint_list(self):
+        """Publish the current waypoint list for visualization in GUI.
+
+        Converts waypoints to ROS message format.
+        """
+        msg = WaypointList()
+        msg.waypoints = [Point(x=wp[0], y=wp[1], z=wp[2]) for wp in self.waypoints]
+        self.waypoint_list_pub.publish(msg)
+
+    def odom_callback(self, msg: Odometry):
+        """Process odometry updates and trigger guidance calculations.
+
+        Updates vehicle state and triggers guidance calculations if a navigation
+        goal is active.
+
+        Args:
+            msg (Odometry): ROS odometry message containing vehicle pose
+        """
+        try:
+            # Extract orientation quaternion to Euler angles
+            orientation_q = msg.pose.pose.orientation
+            roll, pitch, yaw = quat2euler(
+                [orientation_q.w, orientation_q.x, orientation_q.y, orientation_q.z]
+            )
+
+            # Update vehicle state
+            self.state.x = msg.pose.pose.position.x
+            self.state.y = msg.pose.pose.position.y
+            self.state.z = msg.pose.pose.position.z
+            self.state.yaw = yaw
+            self.state.pitch = pitch
+
+            # Initialize filter on first callback
+            if not hasattr(self, 'filter_initialized'):
```

Comment:
You might just want to set `self.filter_initialized = False` in the init(), then let this check be `if not self.filter_initialized`